src/devices/src/virtio/balloon/device.rs

```diff
@@ -0,0 +1,920 @@
+            DeviceState::Activated(ref mem) => mem,
+            // This should never happen, it's been already validated in the event handler.
+            DeviceState::Inactive => unreachable!(),
+        };
+        self.stats_timer.read();
+
+        // The communication is driven by the device by using the buffer
+        // and sending a used buffer notification
+        if let Some(index) = self.stats_desc_index.take() {
+            self.queues[STATS_INDEX].add_used(&mem, index, 0);
+            self.signal_used_queue()
+                .unwrap_or_else(report_balloon_event_fail);
+        }
+    }
+
+    pub(crate) fn process_inflate(&mut self) -> std::result::Result<(), DeviceError> {
+        let mem = match self.device_state {
+            DeviceState::Activated(ref mem) => mem,
+            // This should never happen, it's been already validated in the event handler.
+            DeviceState::Inactive => unreachable!(),
+        };
+        METRICS.balloon.inflate_count.inc();
+
+        let queue = &mut self.queues[INFLATE_INDEX];
+        let mut pages = Vec::with_capacity(MAX_PAGES_IN_DESC);
+        let mut needs_interrupt = false;
+
+        while let Some(head) = queue.pop(&mem) {
+            let len = head.len;
+            if !head.is_write_only() && len % SIZE_OF_U32 as u32 == 0 {
+                for index in (0..len).step_by(SIZE_OF_U32) {
+                    let addr = head
+                        .addr
+                        .checked_add(index as u64)
+                        .ok_or(DeviceError::MalformedDescriptor)?;
+
+                    let page_frame_number = mem
+                        .read_obj::<u32>(addr)
+                        .map_err(|_| DeviceError::MalformedDescriptor)?;
+
+                    pages.push(page_frame_number);
+                }
+            }
+
+            // Acknowledge the receipt of the descriptor.
+            // 0 is number of bytes the device has written to memory.
+            queue.add_used(&mem, head.index, 0);
+            needs_interrupt = true;
+        }
+
+        if needs_interrupt {
+            self.signal_used_queue()
+                .unwrap_or_else(report_balloon_event_fail);
+        }
+
+        // Compact pages into ranges.
+        let page_ranges = compact_page_frame_numbers(&mut pages);
+
+        // Remove the page ranges.
+        for (page_frame_number, range_len) in page_ranges {
+            let guest_addr = GuestAddress((page_frame_number as u64) << VIRTIO_BALLOON_PFN_SHIFT);
+
+            match remove_range(
+                &mem,
+                (guest_addr, u64::from(range_len) << VIRTIO_BALLOON_PFN_SHIFT),
+            ) {
+                Ok(_) => continue,
+                Err(e) => {
+                    error!("Error removing memory range: {:?}", e);
+                }
+            };
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn process_deflate_queue(&mut self) -> bool {
+        let mem = match self.device_state {
```

Comment:
Can this be moved to a function that returns the guest memory?

Reply:
Unfortunately it would result in a double mutable borrow, as the reference to the guest memory needs to be mutable.

Reply:
Could be a macro to dedup code 🤷‍♂️

Reply:
Ok, so this was just purely cosmetical. We can postpone cosmetics for a later time.